frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable);
+
+// checkCopyable is true for copyable, false for assignable.
+// Return: First is from const, second is from ref.
+static const std::pair<bool, bool>& getCopyOrAssignableInfoQuery(
+    Context* context, const Type* t, bool checkCopyable) {
+  QUERY_BEGIN(getCopyOrAssignableInfoQuery, context, t, checkCopyable);
+
+  CHPL_ASSERT((t->isClassType() || t->isCompositeType()) &&
+              "should only be getting called for class or composite types");
+
+  // Inspect type for either kind of copyability/assignability.
+  bool fromConst = false;
+  bool fromRef = false;
+  if (auto classTy = t->toClassType()) {
+    if (classTy->decorator().isManaged() &&
+        classTy->manager()->isAnyOwnedType()) {
+      // Owned class types are copyable from ref iff they are nilable.
+      fromRef = classTy->decorator().isNilable();
+    } else {
+      // Class types of other management are copyable from const.
+      fromConst = true;
+    }
+  } else if (auto ct = t->toCompositeType()) {
+    if (auto at = t->toArrayType()) {
+      if (auto eltType = at->eltType().type()) {
+        // Arrays are copyable/assignable if their elements are
+        getCopyOrAssignableInfo(context, eltType, &fromConst, &fromRef,
+                                checkCopyable);
+      }
+    } else if (auto tt = t->toTupleType()) {
+      // Tuples have the minimum copyable/assignable-ness of their elements
+      if (tt->isStarTuple()) {
+        getCopyOrAssignableInfo(context, tt->elementType(0).type(), &fromConst,
+                                &fromRef, checkCopyable);
+      } else {
+        bool allEltsFromConst = true;
+        bool allEltsFromRef = true;
+        for (int i = 0; i < tt->numElements(); i++) {
+          bool thisEltFromConst = false;
+          bool thisEltFromRef = false;
+          getCopyOrAssignableInfo(context, tt->elementType(i).type(),
+                                  &thisEltFromConst, &thisEltFromRef,
+                                  checkCopyable);
+          allEltsFromConst &= thisEltFromConst;
+          allEltsFromRef &= thisEltFromRef;
+        }
+        fromConst = allEltsFromConst;
+        fromRef = allEltsFromRef;
+      }
+    } else {
+      auto ast = parsing::idToAst(context, ct->id());
+      const AttributeGroup* attrs = nullptr;
+      if (ast) attrs = ast->attributeGroup();
+      if (checkCopyable && attrs &&
+          (attrs->hasPragma(PRAGMA_SYNC) || attrs->hasPragma(PRAGMA_SINGLE))) {
+        // Syncs and singles are copyable
+        // This is a special case to preserve deprecated behavior before
+        // sync/single implicit reads are removed. 12/8/23
```

Comment:
I'm uncertain but I think we might be able to remove these implicit reads now? Either way, what you have in this PR is fine.

Reply:
I assume so as they've been deprecated since https://github.com/chapel-lang/chapel/pull/17287, but I don't have any more context so just left this special behavior the same as prod